lib/likelihood.c

```diff
@@ -0,0 +1,199 @@
+    tsk_id_t parent;
+    tsk_id_t child;
+    
+    for (j = 0; j < edges->num_rows; j++) {
+        parent = edges->parent[j];
+        child = edges->child[j];
+        ret += (edges->right[j] - edges->left[j])
+            * (nodes->time[parent] - nodes->time[child]);
+    }
+    return ret;
+}
+
+int
+msp_unnormalised_log_likelihood_mut(tsk_treeseq_t *ts, double theta,
+                                    double *r_lik) {
+    int ret = 1;
+    tsk_id_t it;
+    tsk_size_t j;
+    tsk_mutation_t mut;
+    tsk_size_t num_mutations = tsk_treeseq_get_num_mutations(ts);
+    const double total_material = get_total_material(ts);
+    const double *node_time = ts->tables->nodes.time;
+    double branch_length;
+    double lik;
+    tsk_tree_t tree;
+    tsk_id_t parent;
+    tsk_id_t child;
+    
+    ret = tsk_tree_init(&tree, ts, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    if (theta > 0) {
+        lik = num_mutations * log(total_material * theta) -
+            total_material * theta;
+        for (it = tsk_tree_first(&tree); it == 1; it = tsk_tree_next(&tree)) {
+            if (it < 0) {
+                ret = it;
+                goto out;
+            }
+            for (j = 0; j < tree.sites_length; j++) {
+                mut = tree.sites[j].mutations[0];
+                child = mut.node;
+                parent = tree.parent[child];
+                branch_length = node_time[parent] - node_time[child];
+                child = parent;
+                parent = tree.parent[parent];
+                while (parent != TSK_NULL && tree.left_child[child]
+                        == tree.right_child[child]) {
+                    branch_length += node_time[parent] - node_time[child];
+                    child = parent;
+                    parent = tree.parent[parent];
+                }
+                child = mut.node;
+                while (tree.left_child[child] >= 0 && tree.left_child[child]
+                        == tree.right_child[child]) {
+                    parent = child;
+                    child = tree.left_child[child];
+                    branch_length += node_time[parent] - node_time[child];
+                }
+                lik += log(branch_length / total_material);
+            }
+        }
+    } else if (num_mutations > 0) {
+        lik = -DBL_MAX;
+    } else {
+        lik = 0;
+    }
+    *r_lik = lik;
+out:
+    tsk_tree_free(&tree);
+    return ret;
+}
+
+int
+msp_log_likelihood_arg(tsk_treeseq_t *ts, double rho, double *r_lik) {
+    int ret = 1;
+    tsk_id_t i;
+    tsk_size_t lineages = tsk_treeseq_get_num_samples(ts);
+    double sim_time = 0;
+    double material = lineages * tsk_treeseq_get_sequence_length(ts);
+    double material_in_children;
+    double material_in_parent;
+    double rate;
+    double gap;
+    double lik = 0;
+    tsk_edge_table_t *edges = &ts->tables->edges;
+    tsk_node_table_t *nodes = &ts->tables->nodes;
+    tsk_id_t *first_parent_edge;
```

Comment:
These must be initialised to NULL:

```
tsk_id_t *first_parent_edge = NULL;
```

This is because we conditionally free the memory based on whether it's null or not. If an error occured allocating first parent edge here, we'd currently free ``last_parent_edge`` (which would be junk left on the stack), resulting in a segfault or similar.